### DISCUSSION TITLE
How to load multiple nifti images per subject in ImageDataset ?

### DISCUSSION THREAD

**Discussion by mpsampat** [Project-MONAI/MONAI#2869] [Q&A]
Hi All! 
I have a basic question. 
I have a classification task where I have 4 nifti files per subjects: these are FLAIR, T2w, T1w and T1w_CE. 
I was able to build an ImageDataset with one one nifti file type as input: 
train_ds = ImageDataset(image_files=images, labels=labels, transform=train_transforms)

where image files is 
'/mnt/kg/rsna//train/00000/T1CE_to_SRI.nii.gz',
 '/mnt/kg/rsna//train/00002/T1CE_to_SRI.nii.gz',
 '/mnt/kg/rsna//train/00003/T1CE_to_SRI.nii.gz',
 '/mnt//kg/rsna//train/00005/T1CE_to_SRI.nii.gz',
 '/mnt/kg/rsna//train/00006/T1CE_to_SRI.nii.gz',

However I cannot figure out how to build an ImageDataset with 4 nifti files per subject, 
Could you please help me with this issue ? 
Thanks
Mehul

**Comment by Nic-Ma** [comment]
Hi @mpsampat ,

For this case, you can try to make every 4 files under 1 dir, use this dir as the image path for 1 data.
MONAI will detect it's a folder and stack the 4 images as 4 channels:
https://github.com/Project-MONAI/MONAI/blob/0.6.0/monai/data/image_reader.py#L383
I will try to enhance the doc-string to make it more clear.

Thanks.

**Reply by mpsampat** [reply]
Hi @Nic-Ma ! thanks so much for your quick response . I tried it out today; 
so i made images a list, where each item is a folder (each folder has 4 nifti files)
now when i try this:
train_ds = ImageDataset(image_files=images, labels=labels, transform=train_transforms)

I get the error: 

/home/msampat/miniconda3/envs/jov/lib/python3.7/site-packages/monai/data/image_reader.py", line 189, in read
    raise FileNotFoundError(f"no DICOMs in: {name}.")
FileNotFoundError: no DICOMs in: '/mnt/kg/rsna//train/00006/'

Not sure why it looks for dicom by default ?

**Reply by Nic-Ma** [reply]
Hi @mpsampat ,

Sorry I made a misleading mistake.
Please try again to pass a list of file paths for the 4 images.
for example:
```
images = [
    [
        "/path/image1-patch1.nii",
        "/path/image1-patch2.nii",
        "/path/image1-patch3.nii",
        "/path/image1-patch4.nii",
    ],
    [
        "/path/image2-patch1.nii",
        "/path/image2-patch2.nii",
        "/path/image2-patch3.nii",
        "/path/image2-patch4.nii",
    ],
    ...
]

train_ds = ImageDataset(image_files=images, ...)
```

Thanks.

**Reply by mpsampat** [reply]
thanks very much @Nic-Ma ! I had to make make some minor modifications but works well. 
this is my same data-set with 1 or 2 nifti files.
#one nifti file per subject; 
<class 'torch.Tensor'> torch.Size([2, 1, 240, 240, 155])
two nifti file per subject; 
<class 'torch.Tensor'> torch.Size([2, 2, 240, 240, 155])

**Reply by mpsampat** [reply]
I made a custom Densenet model: 
def get_new_net():
    return DenseNet(
        spatial_dims=3,
        in_channels=1, # this is 1 for 1 nifti and 2 for 2 nifti and so on
        out_channels=2,
        init_features=16,
        growth_rate=16,
        block_config=(4, 4, 4, 4),
    ).to(device)

**Reply by mpsampat** [reply]
lastly I had to make changes to the transforms: 

# transforms for 1 nifti file
train_transforms = Compose([ScaleIntensity(),AddChannel(),EnsureType()])
val_transforms = Compose([ScaleIntensity(),AddChannel(),EnsureType()])

# transforms for 2 or more nifti files 
train_transforms = Compose([ScaleIntensity(),EnsureType()])
val_transforms = Compose([ScaleIntensity(),EnsureType()])

**Reply by Nic-Ma** [reply]
You can use `EnsureChannelFirst()` transform instead of `AddChannel()`, it can support all cases.

Thanks.

**Reply by Nic-Ma** [reply]
I submitted PR: https://github.com/Project-MONAI/MONAI/pull/2876 to enhance the doc-string for multi-files reading.

Thanks.

**Comment by mpsampat** [comment]
@Nic-Ma  can I  read the 4 nifti files and save them as a 4-D nifti instead ?

**Reply by Nic-Ma** [reply]
Yes, the `SaveImage` transform can support it.

Thanks.
